src/Shared/CommunicationsUtilities.cs

```diff
@@ -646,5 +610,10 @@ internal static int GetHandshakeHashCode(string fileVersion)
                 }
             }
         }
+
+        internal static int AvoidEndOfHandshakeSignal(int x)
+        {
+            return x == EndOfHandshakeSignal ? ~x : x;
```

Comment:
This is clever! Is there any possibility that a non-end-of-handshake value that should legitimately be equal to `EndOfHandshakeSignal` could be confused for another value that should legitimately be equal to `~EndOfHandshakeSignal`?

You are effectively projecting the 2^32 values of an int to 2^32-1 values, reserving one for the end-of-handshake marker. Can it not introduce collisions where two different handshakes would look the same on the wire?

Reply:
That's possible, but I think we have to accept that. Most of the values in the handshake (options, fileVersionMajor, fileVersionMinor, etc.) are explicitly non-negative and reasonably small even though they're all ints as far as the Handshake is concerned, so these won't collide in the near future. Also, if I did my math right, the probability of a collision only increases by 2^-63 for each component with a base probability of erroneous collision of 2^-32, which is pretty small.

The bigger worry is if we add a new component that uses all the bits, and it happens to equal the EndOfHandshakeSignal, since then we could have a hang or an erroneous connection. That's what I was trying to avoid.